### PR TITLE
Address blf GzipFile negative seek error

### LINKED ISSUE #1378: GzipFile BLFWriter negative seek error
### Describe the bug

A [GZipFile](https://github.com/python/cpython/blob/b69ba3d7233013ea8cb95a3e30e2338f7f76b359/Lib/gzip.py#L121) object implementation in `cpython` is unable to be sought negatively while in write mode. This is problematic for the `BLFWriter`. Upon initialization, the `BLFWriter` writes a default header to the file, in which the start and stop timestamps are not set. Once the `BLFWriter` is told to [stop](https://github.com/hardbyte/python-can/blob/2da28c1a1c87776618a60218b0b97800cf2deb34/can/io/blf.py#L583-L584), the remaining data in the buffer is flushed to the file and then the header must be updated with the start and stop timestamps. This is where the `GZipFile` object has problems. The location of the writer must be moved to the 0'th position of the file. When in write mode, the offset position [cannot be negatively sought](https://github.com/python/cpython/blob/b69ba3d7233013ea8cb95a3e30e2338f7f76b359/Lib/gzip.py#L384-L385).

##### Socket CAN setup

Enable CAN interface:

`sudo ip link add dev vcan0 type vcan && sudo ip link set up vcan0`

Generate random CAN messages 1 millisecond apart:

`cangen vcan0 -e -I 18EAFF26 -L i -v -v -v -g 1`

##### python-can setup

Enable the `Logger`:

`python -m can.logger -c vcan0 -i socketcan -b 250000 -f file.blf.gz`

The issue can be seen with the `SizedRotatingLogger` as well.

`python -m can.logger -c vcan0 -i socketcan -b 250000 -f file.blf.gz -s 150000`

### Expected behavior

A `file.blf.gz` should be created without error.

### Additional context

OS and version: Kubuntu 22.04
Python version: 3.7.13
python-can version: `pip install git+https://github.com/hardbyte/python-can@2da28c1a1c87776618a60218b0b97800cf2deb34`
python-can interface/s (if applicable):

<details><summary>Traceback and logs</summary>

`Logger` trace back:

```python
(pycan) jackcook@rollo:~$ python -m can.logger -c vcan0 -i socketcan -b 250000 -f file.blf.gz
Connected to SocketcanBus: socketcan channel 'vcan0'
Can Logger (Started on 2022-08-27 10:59:58.127917)
^CTraceback (most recent call last):
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/logger.py", line 268, in <module>
    main()
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/logger.py", line 264, in main
    logger.stop()
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/io/blf.py", line 589, in stop
    self.file.seek(0)
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/gzip.py", line 371, in seek
    raise OSError('Negative seek in write mode')
OSError: Negative seek in write mode
```

`SizedRotatingLogger` trace back:

```python3
(pycan) jackcook@rollo:~$ python -m can.logger -c vcan0 -i socketcan -b 250000 -f file.blf.gz -s 150000
Connected to SocketcanBus: socketcan channel 'vcan0'
Can Logger (Started on 2022-08-27 11:01:30.452316)
Traceback (most recent call last):
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/logger.py", line 259, in main
    logger(msg)
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/listener.py", line 43, in __call__
    self.on_message_received(msg)
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/io/logger.py", line 206, in on_message_received
    self.do_rollover()
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/io/logger.py", line 336, in do_rollover
    self.writer.stop()
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/io/blf.py", line 589, in stop
    self.file.seek(0)
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/gzip.py", line 371, in seek
    raise OSError('Negative seek in write mode')
OSError: Negative seek in write mode

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/logger.py", line 268, in <module>
    main()
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/logger.py", line 264, in main
    logger.stop()
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/io/logger.py", line 241, in stop
    self.writer.stop()
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/site-packages/can/io/blf.py", line 589, in stop
    self.file.seek(0)
  File "/home/jackcook/anaconda3/envs/pycan/lib/python3.7/gzip.py", line 371, in seek
    raise OSError('Negative seek in write mode')
OSError: Negative seek in write mode
```

</details>

### PR DESCRIPTION
**Pull request by j-c-cook**
The current implementation of the `BLFWriter` requires that the file be negatively sought (https://github.com/hardbyte/python-can/pull/1379#discussion_r952598015). The `cpython` implementation does not allow for `GzipFiles` to be negatively sought while in write mode. Modifying `gzip` is outside of the scope of this project. I see two options here:
1. Modify the `BLFWriter` to close, seek then re-open (@christiansandberg says this was previously being done)
2. Inform the user that the `GzipFile` format is not compatible with the `BLFWriter` (later enhancing `python-can` so that zip formats that can be negatively sought are available)

I'm proposing option 2. 

closes #1378

### CODE REVIEW

**Comment by j-c-cook** [comment]
Now that I've tried out other zip formats:
- `gzip` cannot be negatively sought while in write mode
- `bz2` is not seekable in write mode
- `lzma` is not seekable in write mode
https://docs.python.org/3/library/archiving.html

I'm uncertain what should be done about #1385.

File: can/io/blf.py [hunk 1]
```diff
@@ -395,6 +395,11 @@ def __init__(
             speed and compression (currently equivalent to level 6).
         """
         mode = "rb+" if append else "wb"
+        if type(file) == gzip.GzipFile:
+            raise ValueError(
+                f"The BLFWriter is currently incompatible with "
+                f"{gzip.GzipFile.__name__}s. Try using a different zip format."
+            )
```
Comment (j-c-cook) [inline can/io/blf.py:402]: I'm unaware of any zip format that could be made compatible with the current implementation of `BLFWriter`.

**Comment by zariiii9003** [comment]
I would move the check here:
https://github.com/hardbyte/python-can/blob/cd51ec4c07041c7b3a15b999aa4be5ec3829b05e/can/io/logger.py#L107
A comparison of `real_suffix` against a list of incompatible suffixes (all binary formats) should be sufficient.

I'd suggest to release 4.1.0 once this one is merged.

**Comment by j-c-cook** [comment]
@zariiii9003 Okay, I made modifications.

**Comment by j-c-cook** [comment]
@zariiii9003 tests are passing. Ready for your review.

**Review by pierreluctg — COMMENTED** [review]
Hi @j-c-cook, why are we also removing GZip support for the other formats? I was under the impression that this PR was only to remove GZip for the BLF format, wich make sense because already uses compression internally and GZip is not really necessary.

Anything wrong with for example the ASC GZip support?

**Comment by j-c-cook** [comment]
@pierreluctg I've made a mistake here. I'll open up a new PR.

**Comment by j-c-cook** [comment]
@pierreluctg @zariiii9003 This PR is now migrated to https://github.com/hardbyte/python-can/pull/1429.

**Pull request closed without merging**